keps/sig-node/4112-passdown-resources-to-cri/README.md

```diff
@@ -0,0 +1,1107 @@
+resource requests from the PodSpec.
+
+```diff
+ message UpdateContainerResourcesRequest {
+     // ID of the container to update.
+     string container_id = 1;
+     // Resource configuration specific to Linux containers.
+     LinuxContainerResources linux = 2;
+     // Resource configuration specific to Windows containers.
+     WindowsContainerResources windows = 3;
+     // Unstructured key-value map holding arbitrary additional information for
+     // container resources updating. This can be used for specifying experimental
+     // resources to update or other options to use when updating the container.
+     map<string, string> annotations = 4;
++
++    // Kubernetes resource spec of the container
++    KubernetesResources kubernetes_resources = 5;
+ }
+```
+
+### kubelet
+
+Kubelet code is refactored/modified so that all container resources are known
+before sandbox creation. This mainly consists of preparing all mounts (of all
+containers) early.
+
+Kubelet will be be extended to pass down all mounts, devices, CDI devices and
+the unmodified resource requests and limits to the container runtime in all
+related CRI requests, i.e.  RunPodSandbox, CreateContainer and
+UpdateContainerResources.
+
+For example, take a PodSpec:
+
+```yaml
+apiVersion: v1
+kind: Pod
+...
+spec:
+  containers:
+  - name: cnt-1
+    image: k8s.gcr.io/pause
+    resources:
+      requests:
+        cpu: 1
+        memory: 1G
+        example.com/resource: 1
+      limits:
+        cpu: 2
+        memory: 2G
+        example.com/resource: 1
+    volumeMounts:
+    - mountPath: /my-volume
+      name: my-volume
+  volumes:
+  - name: my-volume
+    emptyDir:
+```
+
+Then kubelet will send the following RunPodSandboxRequest when creating the Pod
+(represented here in yaml format):
+
+```yaml
+RunPodSandboxRequest:
+  config:
+  ...
+    podResources:
+      containers:
+      - name: cnt-1
+        kubernetes_resources:
+          requests:
+            cpu: "1"
+            memory: 1G
+            example.com/resource: "1"
+          limits:
+            cpu: "2"
+            memory: 2G
+            example.com/resource: "1"
+        CDI_devices:
+        - name: example.com/resource=CDI-Dev-1
+        mounts:
+        - container_path: /my-volume
+          host_path: /var/lib/kubelet/pods/<pod-uid>/volumes/kubernetes.io~empty-dir/my-volume
+        - container_path: /var/run/secrets/kubernetes.io/serviceaccount
+          host_path: /var/lib/kubelet/pods/<pod-uid>/volumes/kubernetes.io~projected/kube-api-access-4srqm
+          readonly: true
+        - container_path: /dev/termination-log
+          host_path: /var/lib/kubelet/pods/<pod-uid>/containers/cnt-1/<uuid>
+```
+
+Note that all device plugin resources are passed down in the
+`kubernetes_resources` field but this does not contain any properties of the
+device that was actually allocated for the container. However, these properties
+are exposed through the `CDI_devices`, `mounts` and `devices` fields.
```

Comment:
To clarify: Assuming a podspec includes `example.com/resource: 1` for a container, is the intent that the modifications to the `ContainerConfig` from the `ContainerAllocateResponse` are included when creating the POD sandbox in this case? Are the other fields that can be modified relevant at all (container annotations and environement variables)?

Reply:
The assumption is all modificactions to devices, CDI devices and mounts are included. The current understanding is that annotations and env are not relevant at this phase so they are not included. The can be added, of course, if there is a clear use case for that.

Reply:
@elezar?

Reply:
I think I am comfortable with stating that the environment variables are not relevant here, but in the case of the device plugins with CDI, we do allow for setting container annotations with the prefix `cdi.k8s.io/` to specify which devices are to be injected by the runtime. This has been superseded by the addition of the `CDIDevices` field in the case of CDI and we could state that this is the only supported mechanism in this case.

Given that annotations are sometimes (mis)used to develop POC and early access to features before full CRI support, does adding annotations explicitly here make sense?

Reply:
> Given that annotations are sometimes (mis)used to develop POC and early access to features before full CRI support, does adding annotations explicitly here make sense?

Hmm, I'm not sure, might be. I think annotations could be added for sure, if people think they would be needed. I'll mark this as <UNRESOLVED> in the PR for now (in the next update).